### ISSUE TITLE
too much memory required by python with "mmt translate" with a large file

### ISSUE THREAD

**Issue by cattoni** [ModernMT/MMT#236]
When translating a huge file HUGEFILE with the command

`mmt translate --context-file HUGEFILE --batch < HUGEFILE`

the memory required by the python mmt process keeps on growing without limitations.

With a HUGEFILE of size 3.5G (34M sentences, 519M words), the python mmt memory reached 47GB before dying (available RAM was 64GB without swap) and no translations found in output.

The temporary fix I've found is to divide the HUGEFILE in batches of 1K lines and translating them individually.

**Comment by cattoni** [comment]
Additional information: the reason of the problem is not the context file, but the input file to be translated.

I've just tested the command

`mmt translate --context-file HUGEFILE --batch < SMALLFILE`

where the context file is still the HUGEFILE (34M sentences, 519M words) but the input file is now SMALLFILE containing only 3K sentences to be translated.

In this case there are no memory problems, the RAM required by the python mmt process is low and the command succeedes producing the output as expected.

**Comment by AndRossi** [comment]
Hi @cattoni, does the shell shows any kind of output? (e.g. the beginning of the translation from HUGEFILE). 
Could you please send me the file you are trying to translate, or tell me where to find it?
Thanks!

**Comment by cattoni** [comment]
Hi @AndRossi 

I've just put the HUGE file on mmt.rocks:/home/ubuntu/ROLDANO_files/news.2015.de.shuffled
Concerning the ouput, no translations were generated by the command

`mmt translate --context-file HUGEFILE --batch < HUGEFILE`

that actually died after consuming all the available RAM.

Should you need additional info, don't hesitate to contact me -- @davidecaroselli has my email/phone.

**Comment by AndRossi** [comment]
Hi @cattoni , thank you for the file, I'm downloading it right now.
Meanwhile I have encountered a similar issue with file itwac.it in benchmark_1.1. 

In itwac.it I suspect that the problem is caused by the fact that lines are very very long since the beginning of the document. 
The system fails in translating and printing on screen the first line, and goes on to work on the next ones. 
However, since the first line was not printed, the system won't print the next ones too, because it knows that lines must be output in order.

As a consequence all lines and their translations are kept in central memory, consuming all the available RAM.

Do you think this hypotesis is likely? 
Does your hugefile have very long lines as well?

**Comment by AndRossi** [comment]
Oh wait, forget about that.
I found out that MMT puts the entire file in RAM **regardless**.
It's no surprise that RAM gets entirely consumed!

Fixing it shouldn't be too hard, however. 
I hope to give you good news soon :)

**Comment by cattoni** [comment]
Ok, thanks.

By the way I've just analyzed the lines of the HUGE files.

The 10 longest lines (in bytes) contains from 5.9K to 8.7K bytes.
Their indexes are  39811874,  11195782, 10703039, 33886388, 47871929, 27721027, 29315350, 10689856, 48725195 and 13518238.

**Comment by AndRossi** [comment]
Hi @cattoni, 
I implemented a Python function that works as a separate thread and extracts the translations from the results queue while they are being generated by the translation engine. 
I also made sure that the results queue has a limited size (100 entries, so 100 lines). 
This way it should not be possible to overload RAM.

All changes are inside cli/translation.py, so it's enough to replace that file to use the new fixed version.
Does the translation work fine on your machine now?
